Source/DafnyPipeline.Test/Formatter.cs

```diff
@@ -0,0 +1,2842 @@
+  } else {
+    print ""None?!"";
+  }
+}
+");
+    }
+
+    [Fact]
+    public void FormatterWorksForLabels() {
+      var test = @"
+method BreakLabels(s: seq<int>)
+  requires |s| == 1000
+{
+  label A:
+  for i := 0 to 100
+  {
+    label B:
+    label C:
+    for j := 100 downto 0
+    {
+    }
+  }
+}
+method Test() {
+  var a, b, c, d, e;
+  var i := 0;
+  while (i < M)
+  {
+    var j := 0;
+    label InnerHasLabel:
+    while (j < N)
+    {
+    }
+  }
+  label a:
+  while {
+    case true =>
+      for k := 0 to 10
+        invariant k <= 5
+      {
+        if k == 5 {
+          break break;
+        }
+        c := c + 1;
+      }
+  }
+  var i := 0;
+  while i == 1
+    decreases
+      10 - i,
+      1
+      , 2
+  {
+  }
+  while
+    decreases
+      310 - i,
+      31
+      , 32
+  {
+  }
+  label Loop:
+  while
+    decreases 11 - i,
+              12
+            , 23
+  {
+    case i < 10 =>
+      if i == 929 {
+      } else if i < 7 {
+        i := i + 1;
+        continue Loop;
+      } else {
+        b := true;
+        break Loop;
+      }
+      assert false; // unreachable
+      expect false; // unreachable
+  }
+}
+";
+      FormatterWorksFor(test, test);
+    }
+
+    private void FormatterWorksFor(string testCase, string expectedProgramString = null, bool expectNoToken = false, bool reduceBlockiness = true) {
+      BatchErrorReporter reporter = new BatchErrorReporter();
+      var options = DafnyOptions.Create();
+      DafnyOptions.Install(options);
+      var newlineTypes = Enum.GetValues(typeof(Newlines));
+      foreach (Newlines newLinesType in newlineTypes) {
+        currentNewlines = newLinesType;
+        // This formatting test will remove all the spaces at the beginning of the line
+        // and then recompute it. The result should be the same string.
+        var programString = AdjustNewlines(testCase);
+        var programNotIndented = expectedProgramString != null ? programString : indentRegex.Replace(programString, "");
+        var expectedProgram = expectedProgramString != null ? AdjustNewlines(expectedProgramString) : removeTrailingNewlineRegex.Replace(programString, "");
+
+        ModuleDecl module = new LiteralModuleDecl(new DefaultModuleDecl(), null);
+        Microsoft.Dafny.Type.ResetScopes();
+        BuiltIns builtIns = new BuiltIns();
+        Parser.Parse(programNotIndented, "virtual", "virtual", module, builtIns, reporter);
+        var dafnyProgram = new Program("programName", module, builtIns, reporter);
+        if (reporter.ErrorCount > 0) {
+          var error = reporter.AllMessages[ErrorLevel.Error][0];
+          Assert.False(true, $"{error.message}: line {error.token.line} col {error.token.col}");
+        }
+
+        var firstToken = dafnyProgram.GetFirstTopLevelToken();
+        if (firstToken == null && !expectNoToken) {
+          Assert.False(true, "Did not find a first token");
+        }
+
+        //TODO(Mikael) Make sure every token is owned.
```

Comment:
TODO

Reply:
> TODO

I reactivated the test and made it pass. Good that I had it. Actually, the node hierarchy was incomplete, top-level modules such as `module Outer.A { .. }` were not among the children of `Program`